### PR TITLE
Add Unflatten Module

### PR DESCRIPTION
**Pull request by alvgaona**
This PR implements a feature extension discussed in #41516.

I followed this other PR #22245 to add this other module. While I was at it, I also added `extra_repr()` method in `Flatten` which was missing.

I see there are no unit tests for these modules. Should I add those too? If so, what is the best place I should place these?

### CODE REVIEW

**Comment by alvgaona** [comment]
cc @albanD @ezyang would add yourselves as reviewers? I can't edit those.

**Review by albanD — COMMENTED** [review]
Hi,

Thanks for the PR.

I wasn't aware actually that `Tensor.unflaten` only works with named dimensions. My bad.
This is a bit unsatisfying here are the `nn.Unflatten()` layer wouldn't be able to do the opposite of `nn.Flatten()` in many cases (only NCHW inputs would work).

I wonder if we don't want the user to provide `dim` and `unflattened_size` as ints. So that we can get the same general API as unflatten but we don't have to recreate named Tensors in the middle.

One thing I could see as well is to accept as input, either
- dim as string and the second argument is a namedshape.
- dim is a int and the second argument is a tuple of int  or torch.Size

**Comment by alvgaona** [comment]
Thanks for your feedback. No worries about that.

What you've said is very reasonable. So you're saying that you'd like something like this?

```python
    def __init__(self, dim: int, unflattened_size: Size) -> None:
        super(Unflatten, self).__init__()
        self.dim = dim
        self.unflattened_size = unflattened_size

    def forward(self, input: Tensor) -> Tensor:
        return input.unflatten(
            self.dim,
            (('C', self.unflattened_size[0]), ('H', self.unflattened_size[1]), ('W', self.unflattened_size[2]))
        )
```

**Comment by albanD** [comment]
> So you're saying that you'd like something like this?

For the arguments yes. But your forward here would only support specific inputs.
Also this won't allow the user to specify their own named dimensions if they want to use it.

What about something like:
```python
class UnFlatten(nn.Module)
    def __init__(self, dim: Union[int,str], unflattened_size: Union[tuple,Size]) -> None:
        super(Unflatten, self).__init__()
        if isinstance(dim, int):
            self.named = False
            # Make sure the unflattened_size is a tuple of ints
        else:
            self.named = True
            # Make sure the unflattened_size is a tuple of tuple

        self.dim = dim
        self.unflattened_size = unflattened_size

    def forward(self, input: Tensor) -> Tensor:
        if self.named:
            return input.unflatten(self.dim, self.unflattened_size)
        else:
            dim = self.dim
            if self.dim < 0:
                dim += input.ndim()
            inp_size = list(input.size())
            new_size = inp_size[:dim] + self.unflattened_size + inp_size[dim+1:]
            return input.view(new_size)
```

Also can you add a simple test in test/test_nn.py that makes sure that we get errors when wrong inputs are provided and that it does what we expect.

File: torch/nn/modules/flatten.py [hunk 1]
```diff
@@ -31,3 +34,83 @@ def __init__(self, start_dim: int = 1, end_dim: int = -1) -> None:
...
+
+    def extra_repr(self) -> str:
+        return 'start_dim={}, end_dim={}'.format(
+            self.start_dim, self.end_dim
+        )
+
+
+class Unflatten(Module):
+    r"""
+    Unflattens a tensor into another tensor of shape (N, C, H, W). For use with :class:`~nn.Sequential`.
```
Comment (albanD) [inline torch/nn/modules/flatten.py:46]: I think you need to update this part of the doc to mention the two ways to call this.
The two examples below are very good!

File: test/test_nn.py [hunk 2]
```diff
@@ -8326,6 +8326,51 @@ def test_functional_grad_conv(self):
             torch.nn.grad._grad_input_padding(torch.rand(1, 2, 3), [1, 2, 5], (1,), (0,), (3,))
         self.assertEqual(len(w), 1)
 
+    def test_unflatten(self):
+        tensor_input = torch.randn(32, 49152)
```
Comment (albanD) [inline test/test_nn.py:8330]: nit: could you make the input Tensor smaller here? Just to make sure we don't slow down the tests for no reasons.

**Review by albanD — COMMENTED** [review]
This looks good. Just small comments for the doc/test.

**Comment by alvgaona** [comment]
@albanD absolutely!

**Comment by alvgaona** [comment]
> I think you need to update this part of the doc to mention the two ways to call this.
> The two examples below are very good!

Oh right, yes.

> nit: could you make the input Tensor smaller here? Just to make sure we don't slow down the tests for no reasons.

Of course. Absolutely.

**Review by albanD — APPROVED** [review]
Perfect! Thanks for making the change.

**Comment by alvgaona** [comment]
What should we do with that check that failed @albanD?

**Comment by albanD** [comment]
This is just flaky internal tests :'(  I'll take care of it!
